[BE] Replace M_PI with c10::pi constexpr variable

malfet opened the pull request:
Also, get rid of MSVC specific `_USE_MATH_DEFINES`

Test at compile time that c10::pi<double> == M_PI

malfet committed:
[BE] Replace M_PI with at::native::pi constexpr variable

Also, get rid of MSVC specific `_USE_MATH_DEFINES`

albanD commented on aten/src/ATen/native/Math.h:
```diff
@@ -8,9 +8,12 @@
+namespace at {
+namespace native {
+//TODO: Replace me with std::numbers::pi when C++20 becomes available
+template <typename T> constexpr T pi = 3.14159265358979323846L;
```
Do we know the impact of replacing `M_PI` with this value? Could that lead to some precision changes?

malfet replied:
To ensure there are none, I can add .cpp file that include <math.h> and adds `static_assert(at:native::pi<double> == M_PI)`
The motivation is not to pollute global namespace with M_PI define and also prevent build system brittleness on Windows, where including <cmath> before <math.h> would result in M_PI not being defined

albanD replied:
That sounds good.

> To ensure there are none, I can add .cpp file that include <math.h> and adds static_assert(at:native::pi<double> == M_PI)

Would that be brittle on windows? haha 
It would be nice though to have a dummy PR that adds this check and makes sure it passes on all our CI just as a sanity check.

albanD commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -29,6 +29,9 @@
 namespace native {
+// PI constant for bfloat16 according to https://en.wikipedia.org/wiki/Bfloat16_floating-point_format#Special_values
+template<>
+constexpr c10::BFloat16 pi<c10::BFloat16> = c10::BFloat16(0x4049, c10::BFloat16::from_bits());
```
Why is this not in native/Math.h ?

malfet replied:
It should be moved to the header file, but not sure whether `native/Math.h` is the best place, since it does not include BFloat16.h by default, does it?

albanD reviewed (commented):
What is the original motivation for this change?

albanD reviewed (commented):
compilation failures look related right?

malfet commented:
@albanD yes they are, which were mostly due to the fact, that there seem to be no good way of having a specialized constexpr variable (and BFloat16 constructor is not constexpr) But everything is green now.

albanD reviewed (approved):
LGTM thanks for the update

malfet committed:
Move at::native::pi to c10::

malfet committed:
Add specialization for c10::Half

malfet committed:
Workaround NVCC bug

While trying to access constexpr from kernel, NVCC complains that:
```
identifier "c10::pi<   ::c10::Half> " is undefined in device code
```

walterddr reviewed (approved):
looks good. there are several more .cu/cc files in caffe2/operators that uses _USE_MATH_DEFINES for M_PI. Should we address them as well or leave it be?

malfet commented:
@walterddr good point, let me do it in separate PR

fmassa commented on aten/src/ATen/native/Distributions.h:
```diff
@@ -276,8 +277,8 @@ C10_DEVICE static inline scalar_t digamma_one(scalar_t x) {
     // nvcc does not like that
-    additional_summand = -static_cast<accscalar_t>(M_PI) /
-        compat_tan(static_cast<accscalar_t>(M_PI) * x);
+    additional_summand = -c10::pi<scalar_t> /
+        compat_tan(c10::pi<scalar_t> * x);
```
FYI this was casting to `accscalar_t`
